internal/servers/controller/handlers/workers/worker_service.go

```diff
@@ -76,13 +76,27 @@ func (ws *workerServiceServer) Status(ctx context.Context, req *pbs.StatusReques
 		Controllers: controllers,
 	}
 
-	// TODO (jeff, 05-2021): We should possibly list all connections here with
-	// their statuses, and check those against the found connections below. If
-	// any we think should be closed are marked open by the worker, the worker
-	// should be told to close them.
+	// Build a map of controller-side connections and states. We will
+	// use this to compare against reported states to see if any
+	// reconciliation needs to happen.
+	workerConnectionStatuses := make(map[string]session.ConnectionStatus)
+	workerConnections, err := sessRepo.ListConnectionsByServerId(ctx, req.Worker.PrivateId)
+	if err != nil {
+		return &pbs.StatusResponse{}, status.Errorf(codes.Internal, "Error listing connections for worker: %v", err)
+	}
 
-	var foundConns []string
+	for _, conn := range workerConnections {
+		_, states, err := sessRepo.LookupConnection(ctx, conn.PublicId, nil)
```

Comment:
This could potentially be thousands or tens of thousands of DB calls. I think we need a different way to handle this. One might be to have an analogue of LookupSessions that allows a WithSessionIds but instead WithConnectionIds, so we just give that list as part of a single query. Another might be to return the current status for each connection along with the session in the list. Both have pros and cons and probably worth a deeper discussion with the DB experts.

Reply:
Or possibly the current state could be returned as part of ListConnectionsByServerId.

Reply:
:+1: totally fair. Will look at a different way for sure!